### PR TITLE
HTTP-83 Support SQL queries with project pushdown after a lookup join

### LINKED ISSUE #83: Support SQL queries with project pushdown after a lookup join
Enable the project pushdown capablity to make it possible to use the result of a SQL lookup join query in a dynamic query and restrict the number of columns being used.

Example:

```sql
CREATE TABLE Orders (
    proc_time AS PROCTIME(),
    id STRING,
    `row` ROW<`aStringColumn` STRING, `anIntColumn` INT, `aFloatColumn` FLOAT>
) WITH (
  'connector' = 'datagen',
  'rows-per-second' = '1',
  'fields.id.kind' = 'sequence',
  'fields.id.start' = '1',
  "'fields.id.end' = '5'
);

"CREATE TABLE Customers (
    `enrichedInt` INT,
    `enrichedString` STRING,
    `row` ROW<`aStringColumn` STRING, `anIntColumn` INT, `aFloatColumn` FLOAT>
) WITH ("
    'format' = 'json',"
    'lookup-request.format' = 'json',"
    'lookup-request.format.json.fail-on-missing-field' = 'true',"
    'connector' = 'rest-lookup',"
    'lookup-method' = 'POST',"
    'url' = 'http://localhost:9090/client',"
    'gid.connector.http.source.lookup.header.Content-Type' = 'application/json',"
    'asyncPolling' = 'true'"
);

CREATE TEMPORARY VIEW lookupResult AS
    SELECT o.id, o.`row`, c.enrichedInt, c.enrichedString FROM Orders AS o"
    JOIN Customers FOR SYSTEM_TIME AS OF o.proc_time AS c"
    ON ( o.`row` = c.`row` );

SELECT r.id, r.enrichedInt FROM lookupResult r;
```

### PR DESCRIPTION
**Pull request by OlivierZembri**
#### Description

Ability to support SQL queries that add or remove columns on the result of a lookup join query.

Resolves #83 

##### PR Checklist
- [x] Tests added
- [x] [Changelog](CHANGELOG.md) updated

### CODE REVIEW

**Comment by OlivierZembri** [comment]
@kristoffSC Could you please have  look ?

**Comment by davidradl** [comment]
@OlivierZembri Looks good to me. A couple of comments:
When you say "Added support for using the result of a lookup join operation in a subsequent select query that adds
    or removes columns (project pushdown operation)." I was a bit confused by this wording, are we saying " Adding support for projection pushdown, so for lookup joins the projection columns will be honoured. Prior to this fix the projection columns were not honoured, so the output columns (the projections) could not be changed."  

In terms of migration considerations for existing users, would the select be failed or ignored if a projection was attempted. If the projection was ignored then there maybe more or less columns returned to the user after the fix; which would be worth documenting as a side effect / migration consideration.

**Review by AdrianVasiliu — APPROVED** [review]
LGTM, especially in the light of the unit tests and the real-life tests done on our side.

**Comment by OlivierZembri** [comment]
@davidradl 

> In terms of migration considerations for existing users, would the select be failed or ignored if a projection was attempted.

This is an enhancement proposal without which the Flink job fails with an exception
`java.lang.UnsupportedOperationException: No implementation provided for SupportsProjectionPushDown. Please implement SupportsProjectionPushDown#applyProjection(int[][], DataType)
`

There might be some other usage this code change allow to support, but ours is defined in the description for this PR
(ie. https://github.com/getindata/flink-http-connector/issues/83)

**Comment by davidradl** [comment]
> @davidradl
> 
> > In terms of migration considerations for existing users, would the select be failed or ignored if a projection was attempted.
> 
> This is an enhancement proposal without which the Flink job fails with an exception `java.lang.UnsupportedOperationException: No implementation provided for SupportsProjectionPushDown. Please implement SupportsProjectionPushDown#applyProjection(int[][], DataType) `
> 
> There might be some other usage this code change allow to support, but ours is defined in the description for this PR (ie. #83thanks for the confirmation @OlivierZembri . @kristoffSC unless you have any objections LGTM

**Review by davidradl — APPROVED** [review]

**Comment by AdrianVasiliu** [comment]
@kristoffSC It would make our life much easier if this PR could be reviewed by maintainers and merged relatively quickly. 

There's also @davidradl 's https://github.com/getindata/flink-http-connector/pull/87 that we'll also need, but if reviewing/merging it will take more time, would it be possible to release the connector after the merge of the present PR, as this is the first blocker for our current work.

**Comment by grzegorz8** [comment]
> @kristoffSC It would make our life much easier if this PR could be reviewed by maintainers and merged relatively quickly.
> 
> There's also @davidradl 's #87 that we'll also need, but if reviewing/merging it will take more time, would it be possible to release the connector after the merge of the present PR, as this is the first blocker for our current work.

Both PRs approved with a few minor comments. I'll merge them once they are addressed.

File: src/test/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSourceITCaseTest.java
Comment (OlivierZembri) [reply src/test/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSourceITCaseTest.java:324]: Only the two new unit tests `testLookupJoinProjectPushDown()`and `testLookupJoinProjectPushDownNested()` introduced in this files are for testing the projection push down.
The other tests remains unchanged.

Comment (grzegorz8) [reply src/test/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSourceITCaseTest.java:324]: Sorry for the confusion. `ion` in `Projection` was bolded in my previous comment, which is hard to notice.
What I wanted to propose is to rename these two tests to:
testLookupJoinProject**ion**PushDown() and
testLookupJoinProject**ion**PushDownNested()

Comment (OlivierZembri) [reply src/test/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSourceITCaseTest.java]: My bad, I missed the bold characters... 
This should be fixed now.

**Comment by OlivierZembri** [comment]
@grzegorz8 Thanks for merging.
It would be great to have a release 0.13.0 so that we can use these changes.

**Comment by grzegorz8** [comment]
> @grzegorz8 Thanks for merging. It would be great to have a release 0.13.0 so that we can use these changes.

Release 0.13.0 build started: https://github.com/getindata/flink-http-connector/actions/runs/8544953258/job/23412174122

But I forgot to ask if we need to update any section in README.md, especially when it comes to https://github.com/getindata/flink-http-connector/pull/87

**Comment by AdrianVasiliu** [comment]
@grzegorz8 For some reason https://github.com/getindata/flink-http-connector/actions/runs/8544953258/job/23412174122 failed. Could you please look into it?

Regarding the README.md: well, before the last PRs, it didn't cover the details of the extension points, as such I don't see a particular reason to need it covered in the readme now more than before ;-) But yes, of course, in the long run, it would be nice to have a more complete doc. Moreover, maybe one day a default out-of-the-box implementation will provide enough capabilities for handling common use-cases without the need to implement extension points. But this is a matter for future work.

**Comment by kristoffSC** [comment]
@AdrianVasiliu 
It seems the publish build went through eventually 
https://github.com/getindata/flink-http-connector/actions/runs/8544953258/job/23412947418

**Comment by AdrianVasiliu** [comment]
@kristoffSC Yes, it did, thanks!

**Pull request merged**
